### DISCUSSION TITLE
Video Player

### DISCUSSION THREAD

**Discussion by TheSuunny** [AvaloniaUI/Avalonia#10683] [General]
Hi, is there any way to play the video through Avalonia?

I found a project, but it works only on Windows, i would like cross-platform.

**Comment by WamWooWam** [comment]
To do this in Avalonia itself seems not-super-feasible, because you'd have to either include some kind of cross platform media player (VLC?), codecs or libraries (like libavcodec) with Avalonia itself (which is a lot of data most people don't need, and probably wouldn't perform great), or hook the low level media APIs in every single supported operating system, which wouldn't be super easy as far as i can tell?

**Comment by jmacato** [comment]
I think this won't be in our focus for the time-being, as @WamWooWam said, it's a bit tricky to coordinate a sane approach towards media decoding. First off we have a crippling bug for this kind of applications (#2244) because we need to shuttle image data to-&-fro native codecs to managed .NET space and thus will definitely rely with critical time rendering (#2185). Wrappers like [FFMpeg.Autogen](https://github.com/Ruslan-B/FFmpeg.AutoGen) will definitely ease it up but for now it'd have to take a backseat.

**Comment by Sergey-Terekhin** [comment]
May be this project can be ported to Avalonia: https://github.com/unosquare/ffmediaelement - it already has wrappers around ffmpeg and some WPF controls which relatively easy can be ported

**Comment by jmacato** [comment]
@Sergey-Terekhin yes I've been aware of that project even before I got into Avalonia, but there are still some bugs (#2244) that we need to address first to have an acceptable media playback experience. Besides porting that is not for the faint-hearted, trust me coz i tried before and i judged that we're better off learning how it interacts with ffmpeg.autogen instead of porting ffmediaelement outright.

**Comment by shawnallen85** [comment]
https://github.com/videolan/libvlcsharp/pull/173

Started working on this and hopefully it gets merged in. I've tested it on Windows, MacOS, and Linux (only Ubuntu so far) and it works pretty well.

**Comment by simonmurrell** [comment]
@shawnallen85 Is there a way to plug your video extension into an Avalonia UI application?

**Comment by shawnallen85** [comment]
Sorry for the delay -- this is what I have in my code for using the video extension:

In the XAML:

        <Panel>
            <controls:Video MediaPlayer="{Binding MediaPlayer}" HorizontalAlignment="Stretch" VerticalAlignment="Stretch" />
        </Panel>

In the view model:

I have a property for the MediaPlayer (referenced above in the XAML):

public MediaPlayer MediaPlayer { get; }

Then in my functions I control the media player -- that being I start/stop/change the media, etc.

I hope that helps -- let me know.

Happy Thursday!
Shawn

Edit:

I just noticed this, but my code is referencing the control before it was integrated into the LibVLC project officially -- controls:Video might actually be controls:VideoView.

I think the rest is the same.

**Comment by simonmurrell** [comment]
Do you have a simple demo with this in?

> Sorry for the delay -- this is what I have in my code for using the video extension:
> 
> In the XAML:
> 
> ```
>     <Panel>
>         <controls:Video MediaPlayer="{Binding MediaPlayer}" HorizontalAlignment="Stretch" VerticalAlignment="Stretch" />
>     </Panel>
> ```
> 
> In the view model:
> 
> I have a property for the MediaPlayer (referenced above in the XAML):
> 
> public MediaPlayer MediaPlayer { get; }
> 
> Then in my functions I control the media player -- that being I start/stop/change the media, etc.
> 
> I hope that helps -- let me know.
> 
> Happy Thursday!
> Shawn
> 
> Edit:
> 
> I just noticed this, but my code is referencing the control before it was integrated into the LibVLC project officially -- controls:Video might actually be controls:VideoView.
> 
> I think the rest is the same.

**Comment by shawnallen85** [comment]
I don't -- I'll see if I can modify the app I was using this in to use the official control and push it to github in the next day or so.

**Comment by shawnallen85** [comment]
I figured if I didn't knock it out now it wouldn't happen anytime soon :P.

https://github.com/shawnallen85/Homer

**Comment by simonmurrell** [comment]
Thanks. I will give it a whirl!

**Comment by shawnallen85** [comment]
@simonmurrell Any luck?

**Comment by zls3201** [comment]
I success。
you can read

https://github.com/donandren/vlcsharpavalonia/blob/master/src/LibVLCSharp.Avalonia/NativeVideoPresenter.cs

https://github.com/hudec117/Mpv.NET-lib-/blob/master/src/Mpv.NET.WPFExample/MainWindow.xaml.cs


My Code:

*.csproj add line
```
<PackageReference Include="Mpv.NET" Version="1.2.0" />
```

```cs
/// <summary>
    /// Avalonia VideoView for Windows, Linux and Mac.
    /// </summary>
    public class VideoView : NativeControlHost
    {
        private MpvPlayer player;
        IPlatformHandle CreateWin32(IPlatformHandle parent)
        {
            //var control = base.CreateNativeControlCore(parent);
            player = new MpvPlayer(parent.Handle, "win-x64/mpv-1.dll")
            {
                Loop = true,
                Volume = 100
            };
            player.Load("http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4");
            player.Resume();

            return parent;// new PlatformHandle(control.Handle, "HWND");

        }

        void DestroyWin32(IPlatformHandle handle)
        {
            player.Dispose();
            base.DestroyNativeControlCore(handle);
        }

        protected override IPlatformHandle CreateNativeControlCore(IPlatformHandle parent)
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                return CreateWin32(parent);
            return base.CreateNativeControlCore(parent);
        }

        protected override void DestroyNativeControlCore(IPlatformHandle control)
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                DestroyWin32(control);

            base.DestroyNativeControlCore(control);
        }
    }
```
xaml
```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"			 
        xmlns:vlc="clr-namespace:ZTool.Common;assembly=ZTool"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="ZTool.Controls.VideoPlayerView">
	<vlc:VideoView  HorizontalAlignment="Stretch" VerticalAlignment="Stretch" />
</UserControl>

```

**Comment by bbhxwl** [comment]
Do any of you have a simple VLC demo? I want to use it on raspberry pie, otherwise I can only use pyqt5 it, but I'm familiar with C#

**Comment by Symbai** [comment]
@shawnallen85 There is currently no way to draw something on top of the mediaplayer (because it has no content property) but for LibVLCSharp's WPF implementation it is. Any chance this can be done for Avalonia as well?

**Comment by Geektoolkit** [comment]
Seconding this....I'd really really like to be able to draw on top of the video. Right now event trying to put something on top of it in Z order isn't working (I'm not using full screen mode).  I'd really like to be able to have info about the video thats playing show up in an avalonia window I control while the videos playing

**Comment by radiolondra** [comment]
For anyone is interested:

I created an open source repo on Github which tries to overcome some Avalonia LibVlcSharp issues and missed things.
Give a look to **YAMP** (**Y**et **A**nother **M**edia **P**layer) [here](https://github.com/radiolondra/YAMP).
I created also the setups (x86 & x64) to install YAMP and test it directly (version 0.1.10.220) [here](https://github.com/radiolondra/YAMP/releases).

It is obviously just a starting point. A lot of work is still needed to have YAMP stable.
Any help, contribution, discussion, improvement ideas or just one coffee together is welcome.
Many thanks for your attention.

**Comment by Symbai** [comment]
> I created an open source repo on Github which tries to overcome some Avalonia LibVlcSharp issues and missed things.

As far as I've seen on your project description it does not overcome any of the mentioned issues. It still needs an own window and you still cannot draw elements on top of it. Besides, it is not a control but an application which isn't really what people here are looking for (although they could learn/read the source code). Still thanks for sharing.

**Comment by radiolondra** [comment]
@Symbai Yes it doesn't solve the issues, just tries to live with them and have a decent result.

**Comment by Geektoolkit** [comment]
I've spent quite a few hours on this working on a raspberry pi, and here's what I've found (This is all raspberry pi specific, but I find it interesting because it may help others learn what I learned..which is that video is insanely complex)
1) Even VLC, which is the preferred player for a rapsberry pi bullseye version of Debian...has problems with overlaying...simply turning on subtitles or placing a window on top of it when it's not in full screen mode causes issue
2) Creating a window in avalonia and then calling 'activate' will bring it over the top of vlc, but again due to point 1...it causes it to perform horribly
3) If VLC is in full screen mode, nothing can be placed on top of it (avalonia or otherwise), but it performs great 
4) I and my team tried working around this by using the logo feature of VLC, and that cause perf to drop also
5)  Launching VLC 'windowed' and then trying to resize it using width/height command line args doesn't work (7 year old bug they say is fixed in 4.0...but I don't see an arm build and it's in alpha)
6) I tried using MPV and FFPlay...by using process.start and the correct command lines you can full screen launch them, but they also either fall over when a window is placed over them or aren't GPU accelerated on a pi (I hear I can build a version of them, but then there are reports bullseye broke it, etc)
7) To make windows transparent on debian using compmgr, use -o.0   when running it and then avalonia transparency works, otherwise you get white backgrounds (this is how you can get a window on top of a video

Note: On windows...this all works flawlessly for me. No perf issues with avalonia or vlc, and you can easily put a window on top of the video using Activate, with it transparent and everything working.  

I've tried the VLCcontrol (which I'll continue to use...it's very nice) for Avalonia built upon libvlc, as well as tried shell executiong VLC.

Those are my findings thought they'd be useful for others.  Radiondra if your MPV solution works on arm/debian I could give it a shot, but unless you're using a hardware accleerated version of MPV it wont do me or others much good that are doing this on arm/debian platforms like the pi4

**Comment by radiolondra** [comment]
@Geektoolkit YAMP doesn't use MPV, it uses LibVlc mediaplayer.
You did really an interesting job with Raspberry. I'm working with a couple of pi4 for a TinyML project and peraphs I could use one of them to try Vlc&avalonia things there and test. I'll let you know. The only variable against this is time, which is too short. Anyway, I'll try for sure.

**Comment by radiolondra** [comment]
I created 2 repos on Github to show working samples on how to successfully **embed VLC MediaPlayer** into Avalonia [Window](https://github.com/radiolondra/AvaVLCWindow)s and [UserControl](https://github.com/radiolondra/AvaVLCControl)s.

Still trying to find the best way to add a UserControl (i.e. media player buttons) on top of MediaPlayer. (**Edit**: found the way, see my next post)

The simplest and faster way is to handle a new window (something similar has been used in [YAMP](https://github.com/radiolondra/YAMP)), but I desire to find a more flexible and integrated way. 

Help, suggestions and comments are welcome (obviously).
Thanks.

**Comment by radiolondra** [comment]
**How to add a UserControl on top of VLC MediaPlayer**

I modified the Avalonia [Window ](https://github.com/radiolondra/AvaVLCWindow) repository to show how to put a UserControl on top of MediaPlayer (VideoView).

To do this I added a Content property to VideoView which will create its own new top window and add the Content to this window.

You can create the Content as a new UserControl (for example, a MediaPlayer set of controls) and add this UserControl to the VideoView content, like in:

```
<vlcsharp:VideoView
   HorizontalAlignment="Stretch"
   VerticalAlignment="Stretch"
   x:Name="VideoViewer">

   <!-- Content -->
   <Panel
	VerticalAlignment="Bottom"
	Opacity="0.8"
	Background="Gray">
	<controls:PlayerControls/>
   </Panel>
</vlcsharp:VideoView>
```
In this repo I created a user control with 2 buttons: Play and Stop. 
The control position is at bottom and the buttons are shown using PointerEnter/PointerLeave events, with a simple opacity set to 0.8 (Enter) and to 0 (Leave). Too lazy to add some kind of animation to the UserControl.

I'll modify the Avalonia [UserControl](https://github.com/radiolondra/AvaVLCControl) repository too, adding the same feature, even though you could do it by yourselves.

Remember that, even though it works, the code is very basic and could need still a bit of work to be perfect.

**EDIT**: Just updated [UserControl](https://github.com/radiolondra/AvaVLCControl) repository with the same features.

**Comment by radiolondra** [comment]
**EDIT:** 
Successfully tested on **Kubuntu 18.04**

![AvaVLCWindow-Kubuntux64](https://user-images.githubusercontent.com/20070559/161396810-966f81d9-311d-4c65-8f28-6da3a67988d2.png)

Successfully tested on **MacOS 10.13 (High Sierra)**

![AvaVLCWindow-MacOS](https://user-images.githubusercontent.com/20070559/161376137-ade56fd3-c2ff-4199-aeba-a2bdc0fc4474.png)

**Comment by radiolondra** [comment]
**Test on Raspberry Pi (Raspbian Stretch)**
926MB RAM

As @Geektoolkit said, generated on-top window has no transparency and opacity is not respected. Like @Geektoolkit, I tried every possible combinations of Background and TransparencyLevelHint without any success. Maybe Avalonia has some issues for linux-arm or maybe there are some other settings I don't know.

![AvaVLCWindow-RaspberryPi](https://user-images.githubusercontent.com/20070559/161395884-05fcd867-5527-459e-8cd7-198ff141b17a.png)

Having 926MB of RAM only and just 1 CPU, the video is played in jerks. Anyway, it works BUT the transparency and opacity issues.

**Comment by Geektoolkit** [comment]
To get opacity try xcommgr -o.0 to see if that helps with opacity. I got the pi to work using that command/setting

**Comment by SilkyFowl** [comment]
@radiolondra 
I can confirm that this approach works [just fine with Avalonia.FuncUI(F#)!](https://github.com/SilkyFowl/FuncUI.VideoPlayer/blob/master/src/VideoPlayer/VideoPlayer.fs) Thanks for sharing the idea!

**Comment by radiolondra** [comment]
## New test on Raspberry Pi (DietPi with MATE or Xfce)

### It works fine! 
**Transparency and Opacity are fully honored.**

In my opinion, the problem is not Avalonia vs Raspberry, the problem is what is installed or not in Raspberry.

With **DietPi OS and MATE (or Xfce) as X**, the sample works perfectly. And playback is smooth without jerks, interruptions or whatever.

As root, I installed `libx11-dev, libvlc-dev and vlc`, copied the published sample (built in Windows 10 with `dotnet publish -c release -f net5.0 -r linux-arm`), `chmod 777 the whole folder content` and ran the sample. **It has to be ran as root or with sudo**.

![rpi](https://user-images.githubusercontent.com/20070559/161716907-ab810f65-e1d6-4199-b931-0656faf649c8.png)
![rpi_000](https://user-images.githubusercontent.com/20070559/161716966-e66fee8f-d82b-47fd-9366-4d7da1130265.png)

**Comment by Geektoolkit** [comment]
OMG this would be a game changer for those of us doing video on raspberry pis!  I'll give this a try asap...that looks amazing!  And from what I'm gathering, the video is playing without stuttering?!

**Comment by radiolondra** [comment]
> OMG this would be a game changer for those of us doing video on raspberry pis! I'll give this a try asap...that looks amazing! And from what I'm gathering, the video is playing without stuttering?!

The video is playing per-fec-tly! And my Raspberry is a pi3 model b, so just 1 GB of RAM. But it works like a charme.

DietPi is lighter than whatever else Raspberry OS, but not less complete. Use the 32bits or you'll go against other issues.

Be sure to install MATE or Xfce, I didn't test yet with the other X options.

Be prepared to swallow a couple of shirts before to have it installed in a clean way. I had to flash it 3 times (on the SD). 

At start, it uses the old text installation (it remembers the old RedHat way to install Linux) and it seems not so obvious. After some attempts (you have to manually configure the network too) you will thanks this kind of approach, as it's very powerful.

Remember to use "root" (default password is "dietpi") to install what you need (libx11-dev, libvlc-dev and vlc). Ther's no X pre-installed, so you have to choose which X you want and install it (MATE).

Once X is installed, if you have the default (text) login start MATE with normal "startx" command.

Now copy your published (full) project somewhere (I put mine in  root/Desktop), then add needed libraries above. I don't know if they are the ONLY libs needed, they were for my test project.

You should set exec rights to your project files (I used chmod 777 for all) and start your app as root (or using sudo).

Everything should work. 
**Avalonia now is perfect for Raspberry too.**

**Comment by radiolondra** [comment]
I have just created a pull request on Videolan/LibVLCSharp to include my solution in the distro.
https://github.com/videolan/libvlcsharp/pull/294

**Comment by radiolondra** [comment]
**For anyone is interested**:

A sample to apply in the real world the changes I did in LibVLCSharp.Avalonia

I created **YAMP2** repository (**Y**et-**A**nother-**M**edia-**P**layer v2) [here](https://github.com/radiolondra/YAMP2), which fully uses my new [LibVLCSharp.Avalonia.Unofficial](https://github.com/radiolondra/libvlcsharp/tree/3.x/src/LibVLCSharp.Avalonia.Unofficial) lib.

To make tests, build it by yourselves and run.

Enjoy.

**Comment by radiolondra** [comment]
[The new release](https://github.com/radiolondra/libvlcsharp/tree/3.x/src/LibVLCSharp.Avalonia.Unofficial.UCanvas) of the _Unofficial Avalonia LibVLCSharp_, allows to **display multiple draggable controls over the scene of a video played with LibVLCSharp** in an Avalonia app.

**For anyone is interested.**
Enjoy.

https://user-images.githubusercontent.com/20070559/166648051-296f664c-611c-4607-85d6-ba0e7f2a5ec7.mp4

**Comment by radiolondra** [comment]
## All the links updated (latest libs and samples versions)

### External Links
**VLC video player inside an Avalonia Window/UserControl embedding single static customizable control (e.g.player buttons)**
[LibVLCSharp.Avalonia.Unofficial](https://github.com/radiolondra/libvlcsharp/tree/3.x/src/LibVLCSharp.Avalonia.Unofficial)

**VLC video player inside an Avalonia Window/UserControl embedding multiple draggable customizable controls (e.g.player buttons, images,...)**
[LibVLCSharp.Avalonia.Unofficial.UCanvas](https://github.com/radiolondra/libvlcsharp/tree/3.x/src/LibVLCSharp.Avalonia.Unofficial.UCanvas)

**LibVLCSharp.Avalonia.Unofficial Samples**
[LibVLCSharp.Avalonia.Unofficial.Samples](https://github.com/radiolondra/libvlcsharp/tree/3.x/samples/LibVLCSharp.Avalonia.Unofficial.Samples)

**LibVLCSharp.Avalonia.Unofficial.UCanvas sample**
[LibVLCSharp.Avalonia.Unofficial.UCanvas.Samples](https://github.com/radiolondra/libvlcsharp/tree/3.x/samples/LibVLCSharp.Avalonia.Unofficial.UCanvas.Samples)

**YAMP 2 - Open source video player sample using LibVLCSharp.Avalonia.Unofficial library**
[Github sources](https://github.com/radiolondra/YAMP2)
[**Watch YAMP2 in action**](https://www.screencast.com/t/uTGBnpE6v)

**Comment by xiaokun** [comment]
The overlapped window which hosts the content will be disappeared if it lost focus on macOS 12.6.3. 

I changed some code (add a TextBox) in MainWindow.axaml to produce this problem:
```
<Grid RowDefinitions="*, 40"  ColumnDefinitions="*">		
       <controls:PlayerControl  Name="Player"/>
        <Grid Grid.Row="1" >
            <TextBox  />
        </Grid>
</Grid>
```
the overlapped content will disappear if your click on the TextBox. 

Maybe someone has solution. Thanks.

**Comment by clawton** [comment]
Anyone tried this on a DRM  framebuffer based configuration?

**Comment by timunie** [comment]
I'm going to move this issue into discussions as there is nothing to do on Avalonia code side

**Comment by mysteryx93** [comment]
VLC does not provide the best video quality. MPV is a more powerful player.

I have the [MediaPlayerUI](https://github.com/mysteryx93/MediaPlayerUI.NET) project that allows embedding any audio/video players with a standard UI, however, the MPV implementation hasn't been done yet.

Someone did make some progress in implementing MPV player in Avalonia with OpenGL integration but still has some issues.
https://github.com/mysteryx93/MediaPlayerUI.NET/issues/5

**Reply by hahyes** [reply]
>MPV is a more powerful player.

@mysteryx93 How about performance? I'm curious, because I have very specific case with software used on not very powerful CPU with 4 records playing at the same time. Playing these with WPF and default Media Plaer is a pain (dropping frames, CPU and GPU basically overloaded most of the time), so I'm looking for alternatives.

**Comment by homov** [comment]
[LibMpv.Client](https://github.com/homov/LibMpv) - a complete libmpv API wrapper and some examples of implementing Avalonia controls for displaying video (using OpenGL and Software rendering).

Works on Linux and Windows (maybe MacOS)

**Reply by Sorien** [reply]
Nice to finally see one with proper renderer!

**Reply by mysteryx93** [reply]
Finally someone did it!! About time.

I'll integrate it into MediaPlayerUI when I have the time.

**Reply by timunie** [reply]
@homov 🤯 woah looks promissing. Maybe you want to add it to awsome-avalonia?

**Reply by Symbai** [reply]
Its missing LOTS of documentation and the example is not only extremely limited but also looks really ugly. I mean having to invoke commands in viewmodel using strings, not exactly MVVM friendly and not even C# like (classes, properties etc). It looks like RAW access to unmanaged stuff. How to get/set volume? Video length? Position? Step frame forward/backward? All the common stuff.

More a demonstration project that it can be done rather than something ready to use.

**Reply by homov** [reply]
@Symbai  If you read the post, this is an example of how libmpv can be used with Avalonia. This is not a complete player. But this is the way how one can build it.

**Reply by Symbai** [reply]
I read the post, it still doesn't change much of what I wrote.

**Reply by mysteryx93** [reply]
Could copy some API code from 
https://github.com/hudec117/Mpv.NET-lib-

As for a decent UI, that's what [MediaPlayerUI](https://github.com/mysteryx93/MediaPlayerUI.NET) is for. Needs to be integrated. Avalonia version hasn't yet been completed for video support, including full screen mode that need to be ported from WPF.

**Reply by jack775544** [reply]
This project looks very cool. Glad to see someone finally made something like this.

**Reply by homov** [reply]
Just for information, maybe it will be useful for someone. 
What works (MacOS has not been tested, but it looks like it should work):
- openGL rendering (Windows, Linux, maybe MacOS)
- software rendering (Windows, Linux, Android, should be on MacOS)
- native rendering (Windows, Android - with GPU acceleration)

**Comment by MikeCodesDotNET** [comment]
We've considered creating a video player as a "pro control"; we'd provide commercial support and charge a nominal fee to license it. 

Is this something folks would be interested in?

**Reply by Symbai** [reply]
Depends on the license I would say. If its free to use for non commercial projects I would be interested in. If its behind a paywall regardless, I couldn't care less about it. But do you really believe that there are enough people willing to pay the fees for it that its worth the effort? The video player controls I've seen are always free to use and commercial video players are using self made stuff.

**Reply by mysteryx93** [reply]
I think that MPV + MediaPlayerUI will do a good job already, just need to finish the integration. Custom visual themes can be added easily.

Not sure what a pro control would bring in addition to it.

**Reply by IsaacMarovitz** [reply]
Please please please don't go down the 'freemium' route

**Reply by MikeCodesDotNET** [reply]
> Please please please don't go down the 'freemium' route

We’re exploring lots of options. What would be your problem with charging for additional value adds?

**Reply by EliEron** [reply]
The problem is that it is an extremely slippery slope. How do you decide what is a "value add" control and what is a "core" control? The more useful a feature is the more incentivized you will be to make it paid. Which quickly makes the core part of the framework stale  and restrictive.

I get that you need to make money somehow, but charging for what can easily be argued to be core parts of the framework (a video player is a pretty basic component of a UI framework) is a terrible way to go about it if you want your framework to be useful to non-commercial projects.

Note that this complaint is specifically about the "nominal fee to license it" part of your comment. I have no problem with charging for commercial support. As that is an optional feature that is mostly going to appeal to corporations that can afford it anyway.

**Reply by mysteryx93** [reply]
The monetization of open source software is a very real dilemma. I published a series of cross-platform apps, and am struggling myself to find how to make it sustainable to develop more. Ended up adding ads (my own ads, no data shared to 3rd parties) and an inexpensive license option. Putting any further development on pause until there is some income to justify it, or at least cover the expenses at a minimum.

I think that this is a dilemma for many. And how many projects adopt the abandonware model. What is a good guideline to follow?

Very often you have a job or paid activity, and develop something free on the side. But when that something becomes more serious, then at some point it becomes a problem.

And here I'm looking at the insane amount of work that goes into developing Avalonia 11.

**Reply by Symbai** [reply]
> The monetization of open source software is a very real dilemma

The dilemma is the expectation of open source authors who feel like they should receive something in return for their work and open source users who feel like they can report something and someone else has to work on it. Here is the plain and simple fact: If you don't ask for money, you don't owe people work. Which means YOU decide what you do, when you do. If you think like that and publish open source projects then everything is cool. If you have the feeling it takes too much free time then just stop working on it, still cool.

But as soon as you asked for money, whether via donation or freemium or ads things get worse. Because THEN you definitely owe people work and support and all that stuff. And the real dilemma begins if you make money from it where other people around the world have contributed to for free. Because that is a real bad manner and as a non-lawyer I would question if this is even legal.

Right now as far as i know you pay Avalonia team and receive support in return. This feels, to me as a user, as something external. I mean I can also go to a freelancer site and pay someone else helping me out. This doesn't effect the work here. With that said, I don't even mind if Avalonia is making some controls being paid only if these controls are not standard controls. A good example is a ribbon menu you can find in office software. Such a control is almost only used by business software. A video player control is also very specific. I still won't pay for it but this is nothing a standard application needs. A bad example would be, i dont know, an image button or more features on a Textbox control or that kind of stuff. Things that most applications would need.

With that being said at the end of the day, its the authors decision only. You can put ads in your software. Avalonia can become freemium if its wants to. But there is definitely not a problem with money on open source projects, because they were never intended to make money. Its only some authors, after some time, noticing that things become too large. Yet its up to them to handle that and there are other ways than money as I described above.

**Reply by MikeCodesDotNET** [reply]
This needs a broader discussion, so I've created a [new thread](https://github.com/AvaloniaUI/Avalonia/discussions/11279).

**Reply by seanstardev** [reply]
Making cross-platform video player capability commercial is an understandable temptation, but it is a myopic choice. Getting basic cross-platform 3D, video playback and game controller input should be part of the core package. No-one is more enthusiastic about Avalonia than I - and I fully endorse the commercial bent for advanced components being not-free - but don't cut your nose off to spite your face. Avalonia needs USPs. Cross-platform video for all will catapult your enterprise. Avalonia will be the only game in town, and your commercial endeavours will only benefit.

**Reply by seanstardev** [reply]
For context - I would love to demonstrate cross-platform video in a multi-target, GitHub project app base (web, desktop, Android). It would make it much easier to sell Avalonia to my colleagues on that basis. After that, you can put your own figures on cheques!

**Comment by MikeCodesDotNET** [comment]
If we put everything in the core library, there is nothing to sell; it'd all be FOSS, so I don't know what you'd be selling your colleagues. Users != Revenue.

To be clear, we're not working on video players, 3D or anything else you mentioned. We're focused on XPF and v11 right now.

**Reply by seanstardev** [reply]
Understood. Thanks for reading and keep up the great work :)

**Comment by homov** [comment]
[LibMpv.Client](https://github.com/homov/LibMpv) - video control (a little more progress)

**Comment by jonmdev** [comment]
I need video playback within Avalonia that is effective for embedded videos and streaming videos and can have other visual elements (like text or UI graphics) over top. It must work in Windows/iOS/Android.

Does anyone know if this is currently possible? Or is this not currently possible? 

Perhaps via: 

LibVLCSharp as per: https://stackoverflow.com/questions/75493300/avalonia-libvlcsharp-support-for-ios-android-and-wasm

And:

https://github.com/radiolondra/libvlcsharp/tree/3.x/src/LibVLCSharp.Avalonia.Unofficial

To get the overlay functions working?

Thanks

**Reply by homov** [reply]
Briefly, there is currently no easy way for all the platforms you listed. Simply displaying the video (without controls above it) is entirely possible. The solution for LibVLC works for desktop platforms, but does not work for mobile ones.

One of the possible solutions that you can try is using LibMpv.Client (displaying video using OpenGL FB). This is not an Avalonia control, but you will find examples of how to use the library to write such player. Writing a simple video player is a trivial task.

However, expect that video rendering speed on mobile platforms will not be very good and you will have to figure out how to configure LIbMPV to work on different platforms.

**Reply by homov** [reply]
The second possible way is to use Avalonia in a hosted element of another framework (for example MAUI). Then you can use the native video player for your chosen framework and place Avalonia content above it.

**Reply by jonmdev** [reply]
Thanks for the clarification @homov. Maui has been a nightmare, but I suppose it is my nightmare now, and I will need to stick with it.

Avalonia might have been a nice dream but not so practical for me now.

**Reply by timunie** [reply]
@jonmdev and then there is this for ones wanting to have both worlds: https://github.com/AvaloniaUI/AvaloniaMauiHybrid

**Reply by Monsalma** [reply]
I used the [native control host](https://docs.avaloniaui.net/docs/guides/platforms/android/embed-native-views) approach to display video on Windows (LibVLCSharp) and Android (ExoPlayer). I guess the same approach could be used on iOS, but I personally haven't tried it yet. 

By the way, I think MAUI is internally using the same approach - having different media players on different platforms.

More info on the topic in my blog post:
https://monsalma.net/avalonia-ui-native-video-playback-featuring-libvlcsharp-and-exoplayer/
